### DISCUSSION TITLE
os.walk problem

### DISCUSSION THREAD

**Discussion by denise-amiga** [vlang/v#14581] [General]
I am new to v, I have a problem with the following code, can someone guide me on the right path

```v
import os

fn main(){
	mut files := []string{}
	os.walk('/home/denise', fn (p string) {files << p})
	println(files.len)
}
```
error:
```
test01.v:5:41: error: undefined ident: `files`
    3 | fn main(){
    4 |     mut files := []string{}
    5 |     os.walk('/home/denise', fn (p string) {files << p})
      |                                            ~~~~~
    6 |     println(files.len)
    7 | }
```

Thanks in advance

**Comment by JalonSolov** [comment]
There are 2 problems with this...

1st, to make it a closure, you have to tell it which variables from the outer scope should be available in the closure:
```v
import os

fn main(){
	mut files := []string{}
	os.walk('/home/denise', fn [mut files] (p string) {files << p})
	println(files.len)
}
```
However, this still won't do what you want, because closures get a _copy_ of the parameters from the outer scope - modifying it in the closure won't do any good, as the original won't be changed.

It does at least keep you from getting an error.

**Reply by denise-amiga** [reply]
Thanks for the info, I've been looking at the code and it has a specific function for that case
```v
import os

fn main(){
	mut files := []string{cap: 1024}
	os.walk_with_context('/home/denise', &files, fn (mut files []string, p string) {files << p})
	println(files.len)
}
```
